File: src/core/models/construction-group.ts
import { Commodity } from '../enums/commodity.enum';
import { ICommodityRule } from '../interfaces/commodity-rule.interface';
import { IConstructionGroup } from '../interfaces/construction-group.interface';
import { IConstruction } from '../interfaces/construction.interface';
import { Construction } from './construction';

export class ConstructionGroup implements IConstructionGroup {
  name: string;
  noCredit: boolean;
  group: number;
  size: number;
  colour: number;
  costMul: number;
  bulCost: number;
  fireChance: number;
  cost: number;
  tech: number;
  range: number;
  mpsPages: number;
  count: number;
  resourceId: string;
  commodityRuleCount: ICommodityRule[] = [];
  
  private static groupMap: Map<number, IConstructionGroup> = new Map();
  private static resourceMap: Map<string, IConstructionGroup> = new Map();
  
  constructor(
    name: string,
    noCredit: boolean,
    group: number,
    size: number,
    colour: number,
    costMul: number,
    bulCost: number,
    fireChance: number,
    cost: number,
    tech: number,
    range: number,
    mpsPages: number
  ) {
    this.name = name;
    this.noCredit = noCredit;
    this.group = group;
    this.size = size;
    this.colour = colour;
    this.costMul = costMul;
    this.bulCost = bulCost;
    this.fireChance = fireChance;
    this.cost = cost;
    this.tech = tech;
    this.range = range;
    this.mpsPages = mpsPages;
    this.count = 0;
    this.resourceId = '';
    
    // Initialize commodity rules
    for (let stuff = Commodity.STUFF_INIT; stuff < Commodity.STUFF_COUNT; stuff++) {
      this.commodityRuleCount[stuff] = {
        maxload: 0,
        take: false,
        give: false
      };
    }
    
    // Add this construction group to the static maps
    ConstructionGroup.addConstructionGroup(this);
  }
  
  getCosts(): number {
    return this.cost * this.costMul;
  }
  
  isAllowedHere(x: number, y: number, msg: boolean): boolean {
    // Default implementation always returns true
    // Subclasses should override this method to implement specific placement rules
    return true;
  }
  
  placeItem(x: number, y: number): number {
    // Create a new construction and place it at the specified coordinates
    const construction = this.createConstruction();
    construction.place(x, y);
    this.count++;
    return 1; // Success
  }
  
  createConstruction(): IConstruction {
    // Default implementation creates a basic Construction
    // Subclasses should override this method to create specific construction types
    const construction = new Construction();
    construction.constructionGroup = this;
    construction.initialize();
    return construction;
  }
  
  getName(): string {
    return this.name;
  }
  
  // Static methods for managing construction groups
  
  static addConstructionGroup(constructionGroup: IConstructionGroup): void {
    if (!this.groupMap.has(constructionGroup.group)) {
      this.groupMap.set(constructionGroup.group, constructionGroup);
    } else {
      console.log(`Rejecting ${constructionGroup.name} as ${constructionGroup.group} from ConstructionGroup::groupMap`);
    }
  }
  
  static addResourceID(resID: string, constructionGroup: IConstructionGroup): void {
    if (!this.resourceMap.has(resID)) {
      constructionGroup.resourceId = resID;
      this.resourceMap.set(constructionGroup.resourceId, constructionGroup);
      // In a real implementation, we would create a new ResourceGroup here
    } else {
      console.log(`Rejecting ${constructionGroup.name} as ${constructionGroup.resourceId} from ConstructionGroup::resourceMap`);
    }
  }
  
  static clearGroupMap(): void {
    this.groupMap.clear();
  }
  
  static clearResourceMap(): void {
    this.resourceMap.clear();
  }
  
  static getConstructionGroup(group: number): IConstructionGroup | null {
    return this.groupMap.get(group) || null;
  }
  
  static countConstructionGroup(group: number): number {
    return this.groupMap.has(group) ? 1 : 0;
  }
}

File: LinCityCS.SimulationCore/ConstructionGroup.cs
using System;
using System.Collections.Generic;

namespace LinCityCS.SimulationCore
{
    /// <summary>
    /// Base class for construction groups.
    /// </summary>
    public abstract class ConstructionGroup
    {
        private Dictionary<Commodity, CommodityRule> commodityRuleCount = new Dictionary<Commodity, CommodityRule>();

        /// <summary>
        /// Gets or sets the name of the construction group.
        /// </summary>
        public string Name { get; set; }

        /// <summary>
        /// Gets or sets a value indicating whether the construction group requires credit.
        /// </summary>
        public bool NoCredit { get; set; }

        /// <summary>
        /// Gets or sets the group number.
        /// </summary>
        public int Group { get; set; }

        /// <summary>
        /// Gets or sets the size of the construction.
        /// </summary>
        public int Size { get; set; }

        /// <summary>
        /// Gets or sets the color of the construction.
        /// </summary>
        public int Colour { get; set; }

        /// <summary>
        /// Gets or sets the cost multiplier of the construction.
        /// </summary>
        public int CostMul { get; set; }

        /// <summary>
        /// Gets or sets the bulldoze cost of the construction.
        /// </summary>
        public int BulCost { get; set; }

        /// <summary>
        /// Gets or sets the fire chance of the construction.
        /// </summary>
        public int FireChance { get; set; }

        /// <summary>
        /// Gets or sets the cost of the construction.
        /// </summary>
        public int Cost { get; set; }

        /// <summary>
        /// Gets or sets the tech level required for the construction.
        /// </summary>
        public int Tech { get; set; }

        /// <summary>
        /// Gets or sets the range of the construction.
        /// </summary>
        public int Range { get; set; }

        /// <summary>
        /// Gets or sets the flags of the construction.
        /// </summary>
        public int Flags { get; set; }

        /// <summary>
        /// Gets the commodity rule count.
        /// </summary>
        public Dictionary<Commodity, CommodityRule> CommodityRuleCount => commodityRuleCount;

        /// <summary>
        /// Initializes a new instance of the ConstructionGroup class.
        /// </summary>
        /// <param name="name">The name of the construction group.</param>
        /// <param name="noCredit">Whether the construction group requires credit.</param>
        /// <param name="group">The group number.</param>
        /// <param name="size">The size of the construction.</param>
        /// <param name="colour">The color of the construction.</param>
        /// <param name="costMul">The cost multiplier of the construction.</param>
        /// <param name="bulCost">The bulldoze cost of the construction.</param>
        /// <param name="fireChance">The fire chance of the construction.</param>
        /// <param name="cost">The cost of the construction.</param>
        /// <param name="tech">The tech level required for the construction.</param>
        /// <param name="range">The range of the construction.</param>
        /// <param name="flags">The flags of the construction.</param>
        protected ConstructionGroup(
            string name,
            bool noCredit,
            int group,
            int size,
            int colour,
            int costMul,
            int bulCost,
            int fireChance,
            int cost,
            int tech,
            int range,
            int flags)
        {
            Name = name;
            NoCredit = noCredit;
            Group = group;
            Size = size;
            Colour = colour;
            CostMul = costMul;
            BulCost = bulCost;
            FireChance = fireChance;
            Cost = cost;
            Tech = tech;
            Range = range;
            Flags = flags;

            // Initialize commodity rules
            foreach (Commodity commodity in Enum.GetValues(typeof(Commodity)))
            {
                commodityRuleCount[commodity] = new CommodityRule();
            }
        }

        /// <summary>
        /// Creates a construction of this group.
        /// </summary>
        /// <returns>A new construction of this group.</returns>
        public abstract Construction CreateConstruction();
    }
}
